### DISCUSSION TITLE
Access & Refresh Token Strategy

### DISCUSSION THREAD

**Discussion by Gareth064** [etsy/open-api#542] [Q&A] [answered]
Hi Guys, I am new the the whole OAuth2.0 thing and having to manage refresh tokens, so I'm looking for some advice on how to handle this in respect to Etsy implementation.

The API documentation says the following (Paraphrasing):
`The Access Token lasts for 1 hour` this is what you have to send with any requests to the API.
`The Refresh Token lasts for 24 hours` this is what I have to use to get a new Access Token.

So does this mean that mean that as every hour is approaching, whilst a user is using my app, I need send a refresh token to get the next access token?

Does this also mean if 24 hours passes without a user using my app, that both Access and Refresh Tokens expire and the user will have to go through the Grant Access process again?

I am really struggling to wrap my head around this one.

It feels like I might have to have a background timer service running to constantly check all tokens and going off to refresh them, possibly ever 30 mins to make sure they are refreshed for when the user is going to be using the app again.

**Comment by Gareth064** [comment, marked as answer]
I can't remember where I read that the Refresh Token is only good for 24hrs, but I have just been re-re-re-reading the [docs ](https://developers.etsy.com/documentation/essentials/authentication#step-3-request-an-access-token)and I can see that it says the Refresh Token lasts for 90 days.

![image](https://user-images.githubusercontent.com/22307223/186991622-56c6b7dd-d13e-4d09-807e-78b435f8b45b.png)

That's not as bad any more.

So the strategy is that before I make a call to Etsy's API, on behalf of a customer, I need to check if the Access Token is past it's expiry, and if it is, then call the refresh token endpoint to get a new Access token.

I assume this shouldn't add too much latency to the whole process.

**Comment by dave-shawley** [comment]
Also be aware that refresh tokens can be revoked by user actions within etsy.  You may receive the following response from attempting a token refresh:
```
{
  “error”: “invalid_grant”,
  “error_description”: “refresh_token is revoked”
}
```
If you do, then the access & refresh tokens are dead and you need to re-establish the OAuth 2 connection.
